docs(cn): translate content/docs/hooks-faq.md into Simplified Chinese (#121)
reactjs/zh-hans.reactjs.org: content/docs/hooks-faq.md

---

> id: Hook-faq

Reviewer: 这个地方不用替换吧？

> title: Hook FAQ

Reviewer: ```suggestion
title: Hooks FAQ
```

Reviewer: 英文原文的话，保留

> permalink: docs/Hook-faq.html

Reviewer: 链接会失效的

> prev: Hook-reference.html

Reviewer: 这里也是

* React DOM
* React DOM Server
* React Test Renderer
* React Shallow Renderer

注意，**想要启用 Hooks，所有的 React 包需要 16.8.0 或更高版本**。如果你忘记更新的话 Hooks 是不会起作用的，比如说, React DOM。

React Native 会在它的下一个稳定发行版中全面支持 Hooks。

### 我需要重写我所有的组件吗？ {#do-i-need-to-rewrite-all-my-class-components}

不。我们并 [没有计划](/docs/hooks-intro.html#gradual-adoption-strategy) 从 React 中移除 classes —— 我们都需要不断的发布产品，重写不起。我们推荐在新代码中尝试 Hooks。

### 有什么是 Hooks 能做到而 classes 做不到的？ {#what-can-i-do-with-hooks-that-i-couldnt-with-classes}

Hooks 提供了强大而富有表现力的方式来在组件间复用功能。[「自定义 Hooks」](/docs/hooks-custom.html) 允许我们一瞥能做些什么。[这篇文章](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889) 来自一位 React 核心团队的成员，深入剖析了 Hooks 解锁了哪些新的能力。

### 我的 React 知识还有多少是继续有用的？ {#how-much-of-my-react-knowledge-stays-relevant}

Hooks 是使用你已经知道的 React 特性的一种更直接的方式 —— 比如 state，生命周期，context，以及 refs。它们并没有从根本上改变 React 的工作方式，你对组件，props, 以及自顶向下的数据流的知识并没有改变。

Hooks 确实有它们自己的学习曲线。如果这份文档中遗失了一些什么，[提一个 issue](https://github.com/reactjs/reactjs.org/issues/new)，我们会尽可能地帮你。

### 我应该使用 Hooks，classes，还是两者混用？ {#should-i-use-hooks-classes-or-a-mix-of-both}

当你准备好了，我们鼓励你在新写的组件里开始尝试 Hooks。请确保你团队中的每个人都都愿意使用它们并且熟悉这份文档中的内容。我们不推荐用 Hooks 重写你已有的 classes，除非你本就打算重写它们。（例如：为了修复问题）。

> 你不能用 Hooks *替代* 一个 class 组件，但你绝对可以在一棵树里混用 classed 和使用了 Hooks 的函数组件。一个组件到底是一个 class 还是一个使用了 Hooks 的函数是哪个组件的实现细节。长远来看，我们期望 Hooks 能够成为人们编写 React 组件的主要方式。

Reviewer 2: inside not instead
```suggestion
你不能在一个 class 组件*中*使用 Hooks，但毫无疑问你可以在组件树里混合使用 class 组件和使用了 Hooks 的函数组件。不论一个组件是一个 class 还是一个使用了 Hooks 的函数，都只是这个组件的实现细节而已。长远来看，我们期望 Hooks 能够成为人们编写 React 组件的主要方式。
```

  ).join('\n')
-->

* **[采纳策略](#adoption-strategy)**

>   * [哪个版本的 React 包含了 Hook？](#which-versions-of-react-include-Hook)

Reviewer: ```suggestion
  * [哪个版本的 React 包含了 Hook？](#which-versions-of-react-include-hooks)
```

Reviewer: 这一类的请换回去，谢谢请不要使用批量替换，会导致整个链接失效

  * [在依赖列表中省略函数是否安全？](#is-it-safe-to-omit-functions-from-the-list-of-dependencies)
  * [如果我的 effect 的依赖频繁变化，我该怎么办？](#what-can-i-do-if-my-effect-dependencies-change-too-often)
  * [我该如何实现 shouldComponentUpdate？](#how-do-i-implement-shouldcomponentupdate)
  * [如何记忆计算结果？](#how-to-memoize-calculations)
  * [如何惰性创建昂贵的对象？](#how-to-create-expensive-objects-lazily)
  * [Hooks 是否会因为在渲染时创建函数而显得慢？](#are-hooks-slow-because-of-creating-functions-in-render)
  * [如何避免向下传递回调？](#how-to-avoid-passing-callbacks-down)
  * [如何从 useCallback 读取一个经常变化的值？](#how-to-read-an-often-changing-value-from-usecallback)
* **[底层原理](#under-the-hood)**
  * [React 是如何把对 Hook 的调用和组件联系起来的？](#how-does-react-associate-hook-calls-with-components)
  * [Hooks 使用了哪些现有技术？](#what-is-the-prior-art-for-hooks)

## 采纳策略 {#adoption-strategy}

### 哪个版本的 React 包含了 Hooks？ {#which-versions-of-react-include-hooks}

从 16.8.0 开始，React 在以下模块中包含了 React Hooks 的稳定实现：

* React DOM
* React DOM Server
* React Test Renderer
* React Shallow Renderer

注意，**想要启用 Hooks，所有的 React 包需要 16.8.0 或更高版本**。如果你忘记更新的话 Hooks 是不会起作用的，比如说, React DOM。

React Native 会在它的下一个稳定发行版中全面支持 Hooks。

### 我需要重写我所有的组件吗？ {#do-i-need-to-rewrite-all-my-class-components}

> 不。我们并 [没有计划](/docs/hooks-intro.html#gradual-adoption-strategy) 从 React 中移除 classes —— 我们都需要不断的发布产品，重写不起。我们推荐在新代码中尝试 Hooks。

Author: 这一条我保留意见。

「我们都需要不断地发布产品」，原文中的 all 我个人理解应该也包含了「React 团队也需要不断发布 React 这款产品」的意思，而不仅仅指开发者使用 React 开发出的产品，这里「都」的意思应该得到保留。

「推荐」比「建议」更符合 recommend 的本意，这个应该没有很翻译腔吧？

> 不。我们并 [没有计划](/docs/hooks-intro.html#gradual-adoption-strategy) 从 React 中移除 classes —— 我们都需要不断的发布产品，重写不起。我们推荐在新代码中尝试 Hooks。

Author: 超过 10 天没有反对意见，我就默认通过先关闭了，如有异议再开吧。

  * [const [thing, setThing] = useState() 是什么意思？](#what-does-const-thing-setthing--usestate-mean)
* **[性能优化](#performance-optimizations)**
  * [我可以在更新时跳过 effect 吗？](#can-i-skip-an-effect-on-updates)
  * [在依赖列表中省略函数是否安全？](#is-it-safe-to-omit-functions-from-the-list-of-dependencies)
  * [如果我的 effect 的依赖频繁变化，我该怎么办？](#what-can-i-do-if-my-effect-dependencies-change-too-often)
  * [我该如何实现 shouldComponentUpdate？](#how-do-i-implement-shouldcomponentupdate)
  * [如何记忆计算结果？](#how-to-memoize-calculations)
  * [如何惰性创建昂贵的对象？](#how-to-create-expensive-objects-lazily)
  * [Hook 会因为在渲染时创建函数而变慢吗？](#are-hooks-slow-because-of-creating-functions-in-render)
  * [如何避免向下传递回调？](#how-to-avoid-passing-callbacks-down)
  * [如何从 useCallback 读取一个经常变化的值？](#how-to-read-an-often-changing-value-from-usecallback)
* **[底层原理](#under-the-hood)**
  * [React 是如何把对 Hook 的调用和组件联系起来的？](#how-does-react-associate-hook-calls-with-components)
  * [Hook 使用了哪些现有技术？](#what-is-the-prior-art-for-hooks)

## 采纳策略 {#adoption-strategy}

### 哪个版本的 React 包含了 Hook？ {#which-versions-of-react-include-hooks}

从 16.8.0 开始，React 在以下模块中包含了 React Hook 的稳定实现：

* React DOM
* React DOM Server
* React Test Renderer
* React Shallow Renderer

> 请注意，**要启用 Hook，所有 React 包都必须升级到 16.8.0 或更高版本**。如果你忘记更新诸如 React DOM 之类的包，Hook 将无法运行。

Reviewer: ```suggestion
请注意，**要启用 Hook，所有 React 相关的 package 都必须升级到 16.8.0 或更高版本**。如果你忘记更新诸如 React DOM 之类的 package，Hook 将无法运行。
```

Reviewer: package 不翻译

React Native 会在它的下一个稳定发行版中全面支持 Hook。

### 我需要重写所有的 class 组件吗？ {#do-i-need-to-rewrite-all-my-class-components}

> 不。我们并 [没有计划](/docs/hooks-intro.html#gradual-adoption-strategy) 从 React 中移除 class —— 我们都需要不断地发布产品，重写不起。我们推荐在新代码中尝试 Hook。

Reviewer: ```suggestion
不。我们并 [没有计划](/docs/hooks-intro.html#gradual-adoption-strategy) 从 React 中移除 class —— 我们也需要不断地发布产品，重写成本较高。我们推荐在新代码中尝试 Hook。
```

### 有什么是 Hook 能做而 class 做不到的？ {#what-can-i-do-with-hooks-that-i-couldnt-with-classes}

> Hook 提供了强大而富有表现力的方式来在组件间复用功能。[「自定义 Hook」](/docs/hooks-custom.html) 允许我们一瞥能做些什么。[这篇文章](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889) 来自一位 React 核心团队的成员，深入剖析了 Hook 解锁了哪些新的能力。

Reviewer: ```suggestion
Hook 提供了强大且富有表现力的方式来复用组件间的功能。而[「自定义 Hook」](/docs/hooks-custom.html) 则提供了无限可能。[这篇文章](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889) 来自一位 React 核心团队的成员，深入剖析了 Hook 解锁了哪些新的能力。
```

Reviewer: 感觉一瞥有点奇怪，

Author: 抱歉，不是很理解为什么「a glimpse of what's possible」会译作「无限可能」

Author: Emmm……但「一瞥」是 glimpse 的本意，这一点我觉得应该得以保留。或许可以考虑换个近义词，我考虑下。

Reviewer: 群里有讨论，我刚刚是想意译。
你看看现在这种译法

Reviewer: 我修改了

Author: 通过 [「自定义 Hook」](/docs/hooks-custom.html) 这一节你可以了解能用它做些什么。这篇来自 React 核心团队成员的[文章](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889)则更加深入地剖析了 Hook 解锁了哪些新的能力。

这样如何？个人比较倾向于在不至于太“翻译腔”的前提下，尽可能的还原语言原本的表述方式。

Reviewer: 行~

Reviewer: @tonghuashuo 我只是提供一些参考

### 我的 React 知识还有多少是仍然有用的？ {#how-much-of-my-react-knowledge-stays-relevant}

Hook 是使用你已经知道的 React 特性的一种更直接的方式 —— 比如 state，生命周期，context，以及 refs。它们并没有从根本上改变 React 的工作方式，你对组件，props, 以及自顶向下的数据流的知识并没有改变。

Hook 确实有它们自己的学习曲线。如果这份文档中遗失了一些什么，[提一个 issue](https://github.com/reactjs/reactjs.org/issues/new)，我们会尽可能地帮你。

### 我应该使用 Hook，class，还是两者混用？ {#should-i-use-hooks-classes-or-a-mix-of-both}

当你准备好了，我们鼓励你在写新组件的时候开始尝试 Hook。请确保你团队中的每个人都愿意使用它们并且熟知这份文档中的内容。我们不推荐用 Hook 重写你已有的 class，除非你本就打算重写它们。（例如：为了修复bug）。

你不能在一个 class 组件*中*使用 Hook，但毫无疑问你可以在组件树里混合使用 class 组件和使用了 Hook 的函数组件。不论一个组件是 class 还是一个使用了 Hook 的函数，都只是这个组件的实现细节而已。长远来看，我们期望 Hook 能够成为人们编写 React 组件的主要方式。

### Hook 能否覆盖 class 的所有使用场景？ {#do-hooks-cover-all-use-cases-for-classes}

我们给 Hook 设定的目标是尽早覆盖 class 的所有使用场景。目前暂时还没有对应不常用的 `getSnapshotBeforeUpdate` 和 `componentDidCatch` 生命周期的 Hook 等价写法，但我们计划尽早把它们加进来。

目前 Hook 还处于早期阶段，一些第三方的库可能还暂时无法兼容 Hook。

### Hook 会替代 render props 和高阶组件吗？ {#do-hooks-replace-render-props-and-higher-order-components}

通常，render props 和高阶组件只渲染一个子节点。我们认为让 Hook 来服务这个使用场景更加简单。这两种模式仍有用武之地，（例如，一个虚拟滚动条组件或许会有一个 `renderItem` 属性，或是一个可见的容器组件或许会有它自己的 DOM 结构）。但在大部分场景下，Hook 足够了，并且能够帮助减少嵌套。

> ### Hook 对于 Redux connect() 和 React Router 等流行的 API 来说，意味着什么？ {#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}

Reviewer: ```suggestion
### Hook 对于 Redux `connect()` 和 React Router 等流行的 API 来说，意味着什么？ {#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}
```

-->

* **[采纳策略](#adoption-strategy)**

>   * [哪个版本的 React 包含了 Hook？](#which-versions-of-react-include-Hook)

Reviewer: 链接跳转的标题不要改= =

});
```

> 对 `act()` 的调用也会清空它们内部的 Effects。

Reviewer: ```suggestion
对 `act()` 的调用也会清空它们内部的 effect。
```

}
```

现在假设我们想要编写一些逻辑以便在用户移动鼠标时改变 `left` 和 `top`。注意到我们是如何必须手动把这些字段合并到之前的 state 对象的：

```js{4,5}
// ...
  useEffect(() => {
    function handleWindowMouseMove(e) {
      // 展开 「...state」 以确保我们没有 「丢失」 width 和 height
      setState(state => ({ ...state, left: e.pageX, top: e.pageY }));
    }
    // 注意：这是个简化版的实现
    window.addEventListener('mousemove', handleWindowMouseMove);
    return () => window.removeEventListener('mousemove', handleWindowMouseMove);
  }, []);
  // ...
```

这是因为当我们更新一个 state 变量，我们会 *替换* 它的值。这和 class 中的 `this.setState` 不一样，后者会把更新后的字段 *合并* 入对象中。

> 如果你怀念自动合并，你可以写一个自定义的 `useLegacyState` Hook 来合并对象 state 的更新。然而，相反的**我们推荐吧 state 切分成多个 state 变量，每个变量包含的不同值会在同时发生变化。**

Author: 1）「吧」改过来了，感谢指正。
2）你试试看把「based on」和「which values tend to change together」分开来断句，其实是通的。「which」指代「values」而不是「multiple state variables」。

}
```

> 注意看我们是如何做到把对 `position` 这个 state 变量的 `useState` 调用和相关的 effect 移动到一个自定义 custom Hook 但不改变它们的代码的。如果所有的 state 都存在同一个对象中，想要抽取出来就比较难了。

Author: 2）我把「without changing their code」倒装一下，放到句子前面来；虽然没法把长句缩短，但应该会比之前顺一点。
「注意看我们是如何做到不改动代码就把对 `position` 这个 state 变量的 `useState` 调用和相关的 effect 移动到一个自定义 Hook 的。」

}
```

如果你先点击「Show alert」然后增加计数器的计数，那这个 alert 会 **在你点击『Show alert』按钮时** 显示 `count` 变量。这避免了那些因为假设 props 和 state 没有改变的代码引起问题。

如果你刻意地想要从某些异步回调中读取 *最新的* state，你可以用 [一个 ref](/docs/hooks-faq.html#is-there-something-like-instance-variables) 来保存它，修改它，并从中读取。

> 最后，你看到陈旧的 props 和 state 的另一个可能的原因是，如果你使用了「依赖数组」优化但没有正确地指定所有的依赖。举个例子，如果一个 effect 指定了 `[]` 作为第二个参数，但在内部读取了 `someProp`，它会一直「看到」 `someProp` 的初始值。解决办法是要么移除依赖数组，或者修正它。 这是 [你该如何处理函数](#is-it-safe-to-omit-functions-from-the-list-of-dependencies)，而这是关于如何减少 effect 的运行而不必错误的跳过依赖的 [一些常见策略](#what-can-i-do-if-my-effect-dependencies-change-too-often)。

Author: 断句的问题，我觉得这里应该还好吧。我把最后这一句里两个「这是」稍微意译了一下，自己回过头来看觉得有点翻译腔。

  }
```

可能的话尽量避免这种模式。

### 我可以引用一个函数组件吗？ {#can-i-make-a-ref-to-a-function-component}

尽管你不应该经常需要这么做，但你可以通过 [`useImperativeHandle`](/docs/hooks-reference.html#useimperativehandle) Hook 暴露一些命令式的方法给父组件。

### `const [thing, setThing] = useState()` 是什么意思？ {#what-does-const-thing-setthing--usestate-mean}

如果你不熟悉这个语法，可以查看 State Hook 文档中的 [解释](/docs/hooks-state.html#tip-what-do-square-brackets-mean) 一节。

## 性能优化 {#performance-optimizations}

### 我可以在更新时跳过一个 effect 吗？ {#can-i-skip-an-effect-on-updates}

> 可以的。参见 [条件式的发起一个 effect](/docs/hooks-reference.html#conditionally-firing-an-effect)。注意，忘记处理更新常会 [导致 bugs](/docs/hooks-effect.html#explanation-why-effects-run-on-each-update)，这就是不以此为默认行为的原因。

Reviewer 2: ```suggestion
可以的。参见 [条件式的发起一个 effect](/docs/hooks-reference.html#conditionally-firing-an-effect)。注意，忘记处理更新常会 [导致 bug](/docs/hooks-effect.html#explanation-why-effects-run-on-each-update)，这也正是我们没有默认使用条件式 effect 的原因。
```

### 我该如何实现 `shouldComponentUpdate`? {#how-do-i-implement-shouldcomponentupdate}

你可以用 `React.memo` 包裹一个组件来对它的 props 进行浅比较：

```js
const Button = React.memo((props) => {
  // 你的组件
});
```

> 这不是一个 Hook 因为它的写法和 Hooks 不同。`React.memo` 等效于 `PureComponent`，但它之比较 props。（你也可以通过第二个参数指定一个自定义的比较函数来比较新旧 props。如果函数返回 true，就会跳过更新。）

Reviewer 2: ```suggestion
这不是一个 Hook 因为它的写法和 Hooks 不同。`React.memo` 等效于 `PureComponent`，但它只比较 props。（你也可以通过第二个参数指定一个自定义的比较函数来比较新旧 props。如果函数返回 true，就会跳过更新。）
```

> `React.memo` 不比较 state 因为没有单一的 state 对象可供比较。但你也可以让子节点变为纯组件，或者针织 [用 `useMemo` 优化每一个具体的子节点](/docs/hooks-faq.html#how-to-memoize-calculations)。

Reviewer 2: ```suggestion
`React.memo` 不比较 state 因为没有单一的 state 对象可供比较。但你也可以让子节点变为纯组件，或者 [用 `useMemo` 优化每一个具体的子节点](/docs/hooks-faq.html#how-to-memoize-calculations)。
```

}
```

注意这种方式在循环中是无效的，因为 Hook 调用 [不能](/docs/hooks-rules.html) 被放在循环中。但你可以为列表项抽取一个单独的组件，并在其中调用 `useMemo`。

### 如何惰性创建昂贵的对象？ {#how-to-create-expensive-objects-lazily}

如果输入的内容相同，`useMemo` 允许你 [记住一次昂贵的计算](#how-to-memoize-calculations)。但是，这仅作为一种提示，并不 *保证* 计算不会重新运行。但有时候需要确保一个对象仅被创建一次。

**第一个常见的使用场景是当创建初始 state 很昂贵时：**

```js
function Table(props) {
  // ⚠️ createRows() 每次渲染都会被调用
  const [rows, setRows] = useState(createRows(props.count));
  // ...
}
```

为避免重新创建被忽略的初始 state，我们可以传一个 **函数** 给 `useState`：

```js
function Table(props) {
  // ✅ createRows() 只会被调用一次
  const [rows, setRows] = useState(() => createRows(props.count));
  // ...
}
```

React 只会在首次渲染时调用这个函数。参见 [`useState` API 参考](/docs/hooks-reference.html#usestate)。

**你或许也会偶尔想要避免重新创建 `useRef()` 的初始值。**举个例子，或许你想确保某些命令式的 class 实例只被创建一次：

```js
function Image(props) {
  // ⚠️ IntersectionObserver 在每次渲染都会被创建
  const ref = useRef(new IntersectionObserver(onIntersect));
  // ...
}
```

> `useRef` **不会** 像 `useState` 那样接受一个特殊的函数重载。相反，你可以辨析你自己的函数来创建并将其设为惰性的：

Reviewer 2: ```suggestion
`useRef` **不会** 像 `useState` 那样接受一个特殊的函数重载。相反，你可以编写你自己的函数来创建并将其设为惰性的：
```

}
```

总而言之，从维护的角度来这样看更加方便（不用不断转发回调），同时也避免了回调的问题。像这样向下传递 `dispatch` 是处理深度更新的推荐模式。

注意，你依然可以选择是要把应用的 *state* 作为 props 向下传递（更显明确）还是作为作为 context（对很深的更新而言更加方便）。如果你也使用 context 来向下传递 state，请使用两种不同的 context 类型 —— `dispatch` context 永远不会变，因此组件通过读取它就不需要重新渲染了，除非它们还需要应用的 state。

### 如何从 `useCallback` 读取一个经常变化的值？ {#how-to-read-an-often-changing-value-from-usecallback}

>注意
>
>我们推荐 [在 context 中向下传递 `dispatch`](#how-to-avoid-passing-callbacks-down) 而非在 props 中使用独立的回调。下面的方法仅仅出于文档完整性考虑，以及作为一条出路在此提及。
>

> >同时也请注意这种模式在 [并行模式](/blog/2018/03/27/update-on-async-rendering.html) 下可能会导致一些问题。我们计划在未来提供一个更加符合人体工程学的太呆方案，但当下最安全的解决方案是，如果回调所依赖的值变化了，总是让回调失效。

Reviewer 2: ```suggestion
>同时也请注意这种模式在 [并行模式](/blog/2018/03/27/update-on-async-rendering.html) 下可能会导致一些问题。我们计划在未来提供一个更加合理的替代方案，但当下最安全的解决方案是，如果回调所依赖的值变化了，总是让回调失效。
```

function Example(props) {
  // 把最新的 props 保存在一个 ref 中
  let latestProps = useRef(props);
  useEffect(() => {
    latestProps.current = props;
  });

  useEffect(() => {
    function tick() {
      // 在任何时候读取最新的 props
      console.log(latestProps.current);
    }

    const id = setInterval(tick, 1000);
    return () => clearInterval(id);
  }, []); // 这个 effect 从不会重新执行
}
```

仅当你是在找不到更好办法的时候才这么做，因为依赖于变更会使得组件更难以预测。如果有某些特定的模式无法很好地转化成这样，[发起一个 issue](https://github.com/facebook/react/issues/new) 并配上可运行的实例代码以便，我们会尽可能帮助你。

### 我该如何实现 `shouldComponentUpdate`? {#how-do-i-implement-shouldcomponentupdate}

你可以用 `React.memo` 包裹一个组件来对它的 props 进行浅比较：

```js
const Button = React.memo((props) => {
  // 你的组件
});
```

这不是一个 Hook 因为它的写法和 Hook 不同。`React.memo` 等效于 `PureComponent`，但它只比较 props。（你也可以通过第二个参数指定一个自定义的比较函数来比较新旧 props。如果函数返回 true，就会跳过更新。）

`React.memo` 不比较 state 因为没有单一的 state 对象可供比较。但你也可以让子节点变为纯组件，或者 [用 `useMemo` 优化每一个具体的子节点](/docs/hooks-faq.html#how-to-memoize-calculations)。

### 如何记忆计算结果？ {#how-to-memoize-calculations}

[`useMemo`](/docs/hooks-reference.html#usememo) Hook 允许你通过「记住」上一次计算结果的方式在多次渲染的之间缓存计算结果：

```js
const memoizedValue = useMemo(() => computeExpensiveValue(a, b), [a, b]);
```

> 这行代码会调用 `computeExpensiveValue(a, b)`。但如果输入 `[a, b]` 自上次值以来没有改变过，`useMemo` 会跳过再一次的调用，简单复用它上一次返回的值。

Author: 你觉得「跳过二次调用」如何？我想尽可能保留「skip」的本意，「避免」可能更适合「avoid」。

}
```

> 无论是和，我们都 **不推荐使用这种模式** 并仅处于文档文整形而把它展示在这里。相反的，我们更倾向于 [避免很深的向下传递回调](#how-to-avoid-passing-callbacks-down)。

Reviewer 2: ```suggestion
无论是和，我们都 **不推荐使用这种模式** 并仅为了文档完整性而把它展示在这里。相反的，我们更倾向于 [避免很深的向下传递回调](#how-to-avoid-passing-callbacks-down)。
```

## 底层原理 {#under-the-hood}

### React 是如何把对 Hook 的调用和组件联系起来的？ {#how-does-react-associate-hook-calls-with-components}

React 保持对当先渲染中的组件的追踪。多亏了 [Hooks 规范](/docs/hooks-rules.html)，我们得知 Hooks 只会在 React 组件中被调用（或自定义 Hooks —— 同样只会在 React 组件中被调用）。

每个组件内部都有一个「记忆单元格」列表。它们只不过是我们用来存储一些数据的 JavaScript 对象。当你用 `useState()` 调用一个 Hook 的时候，它会读取当前的单元格（或在首次渲染时将其初始化），让后把指针移动到下一个。这就是多个 `useState()` 调用会得到各自独立的本地 state 的原因。

### Hooks 使用了哪些现有技术？ {#what-is-the-prior-art-for-hooks}

> Hooks 从多个不同的来源合成想法：

Reviewer 2: ```suggestion
Hooks 由不同的来源的多个想法构成：
```

* [react-future](https://github.com/reactjs/react-future/tree/master/07%20-%20Returning%20State) 这个仓库中我们对函数式 APIs 的老旧实验。
* React 社区对 render prop APIs 的实验，包括 [Ryan Florence](https://github.com/ryanflorence) 的 [Reactions Component](https://github.com/reactions/component)。
* [Dominic Gannaway](https://github.com/trueadm) 的用 [`adopt` 关键字](https://gist.github.com/trueadm/17beb64288e30192f3aa29cad0218067) 作为 render props 的语法糖的提案。
* [DisplayScript](http://displayscript.org/introduction.html) 中的 state 变量和 state 单元格。
* ReasonReact 中的 [Reducer components](https://reasonml.github.io/reason-react/docs/en/state-actions-reducer.html)。
* Rx 中的 [Subscriptions](http://reactivex.io/rxjs/class/es6/Subscription.js~Subscription.html)。

> * 多核 OCaml 中的 [Algebraic effects](https://github.com/ocamllabs/ocaml-effects-tutorial#2-effectful-computations-in-a-pure-setting)。

Reviewer 2: ```suggestion
* Multicore OCaml 提到的 [Algebraic effects](https://github.com/ocamllabs/ocaml-effects-tutorial#2-effectful-computations-in-a-pure-setting)。
```

    return newObserver;
  }

  // 当你需要时，调用 getObserver()
  // ...
}
```

> 这避免了创建一个昂贵的对象直到它首次被真正需要。如果你使用 Flow 或 TypeScript，你还可以为了方便给 `getObserver()` 一个不可为 null 的类型。

Reviewer 3: 这避免了在我们首次真正需要一个对象之前就创建它。如果你使用 Flow 或 TypeScript，你还可以为了方便给 `getObserver()` 一个不可为 null 的类型。

- 第一句的翻译还有待商榷；

Author: 「这避免了我们在一个对象被首次真正需要之前就创建它」，这样如何？

### Hook 会因为在渲染时创建函数而变慢吗？ {#are-hooks-slow-because-of-creating-functions-in-render}

不会。在现代浏览器中，闭包和类的原始性能只有在极端场景下才会有明显的差别。

除此之外，可以认为 Hook 的设计在某些方面更加高效：

* Hook 避免了 class 需要的额外开支，像是创建类实例和在构造函数中绑定事件处理器的成本。

* **符合语言习惯的代码在使用 Hook 时不需要很深的组件树嵌套**。这个现象在使用高阶组件、render props、和 context 的代码库中非常普遍。组件树小了，React 的工作量也随之减少。

> 传统上认为，在 React 中使用内联函数对性能的影响与在每次渲染时传递新的回调是如何破坏子组件的 `shouldComponentUpdate` 优化的有关。 Hook 从三个方面解决了这个问题。

Reviewer 3: 传统上认为，React 内联函数对性能的影响、以及在每次渲染时向子组件传递新的回调会如何破坏子组件的 `shouldComponentUpdate` 优化，二者是有关联的。 Hook 从三个方面解决了这个问题。

- 有点绕，待商榷；

Author: 「传统上认为，在 React 中使用内联函数对性能的影响，与每次渲染都传递新的回调会如何破坏子组件的 `shouldComponentUpdate` 优化有关。」，这样如何？

}
```

总而言之，从维护的角度来这样看更加方便（不用不断转发回调），同时也避免了回调的问题。像这样向下传递 `dispatch` 是处理深度更新的推荐模式。

注意，你依然可以选择是要把应用的 *state* 作为 props 向下传递（更显明确）还是作为作为 context（对很深的更新而言更加方便）。如果你也使用 context 来向下传递 state，请使用两种不同的 context 类型 —— `dispatch` context 永远不会变，因此组件通过读取它就不需要重新渲染了，除非它们还需要应用的 state。

### 如何从 `useCallback` 读取一个经常变化的值？ {#how-to-read-an-often-changing-value-from-usecallback}

>注意
>
>我们推荐 [在 context 中向下传递 `dispatch`](#how-to-avoid-passing-callbacks-down) 而非在 props 中使用独立的回调。下面的方法仅仅出于文档完整性考虑，以及作为一条出路在此提及。
>
>同时也请注意这种模式在 [并行模式](/blog/2018/03/27/update-on-async-rendering.html) 下可能会导致一些问题。我们计划在未来提供一个更加合理的替代方案，但当下最安全的解决方案是，如果回调所依赖的值变化了，总是让回调失效。

在某些罕见场景中，你可能会需要用 [`useCallback`](/docs/hooks-reference.html#usecallback) 记住一个回调，但由于内部函数必须经常重新创建，记忆效果不是很好。如果你想要记住的函数是一个事件处理器并且在渲染期间没有被用到，你可以 [把 ref 当做实例变量](#is-there-something-like-instance-variables) 来用，并手动把最后提交的值保存在它当中：

```js{6,10}
function Form() {
  const [text, updateText] = useState('');
  const textRef = useRef();

> <<<<<<< HEAD

Reviewer: 检查下冲突

Author: 好了

}
```

> 无论是和，我们都 **不推荐使用这种模式** 并仅为了文档完整性而把它展示在这里。相反的，我们更倾向于 [避免很深的向下传递回调](#how-to-avoid-passing-callbacks-down)。

Reviewer 3: 无论如何，我们都 **不推荐使用这种模式**，我们把它展示在这里仅为了文档完整性。相反的，我们更倾向于 [避免向下深入传递回调](#how-to-avoid-passing-callbacks-down)。

Author: 「无论如何，我们都 **不推荐使用这种模式** ，只是为了文档的完整性而把它展示在这里。相反的，我们更倾向于 [避免向下深入传递回调](#how-to-avoid-passing-callbacks-down)。」，这样如何？

## 底层原理 {#under-the-hood}

### React 是如何把对 Hook 的调用和组件联系起来的？ {#how-does-react-associate-hook-calls-with-components}

React 保持对当先渲染中的组件的追踪。多亏了 [Hook 规范](/docs/hooks-rules.html)，我们得知 Hook 只会在 React 组件中被调用（或自定义 Hook —— 同样只会在 React 组件中被调用）。

每个组件内部都有一个「记忆单元格」列表。它们只不过是我们用来存储一些数据的 JavaScript 对象。当你用 `useState()` 调用一个 Hook 的时候，它会读取当前的单元格（或在首次渲染时将其初始化），让后把指针移动到下一个。这就是多个 `useState()` 调用会得到各自独立的本地 state 的原因。

### Hook 使用了哪些现有技术？ {#what-is-the-prior-art-for-hooks}

Hook 由不同的来源的多个想法构成：

> * [react-future](https://github.com/reactjs/react-future/tree/master/07%20-%20Returning%20State) 这个仓库中我们对函数式 APIs 的老旧实验。

Reviewer: ```suggestion
* [react-future](https://github.com/reactjs/react-future/tree/master/07%20-%20Returning%20State) 这个仓库中包含我们对函数式 API 的老旧实验。
```

> * React 社区对 render prop APIs 的实验，包括 [Ryan Florence](https://github.com/ryanflorence) 的 [Reactions Component](https://github.com/reactions/component)。

Reviewer: ```suggestion
* React 社区对 render prop API 的实验，包括 [Ryan Florence](https://github.com/ryanflorence) 的 [Reactions Component](https://github.com/reactions/component) 等。
```
